### PR TITLE
Method for codon optimization of sequences

### PR DESCRIPTION
**Pull request by crockeraw**
- [X] I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

- [X] I have read the ``CONTRIBUTING.rst`` file, have run ``pre-commit``
locally, and understand that continuous integration checks will be used to
confirm the Biopython unit tests and style checks pass with these changes.

- [X] I have added my name to the alphabetical contributors listings in the files
``NEWS.rst`` and ``CONTRIB.rst`` as part of this pull request, am listed
already, or do not wish to be listed. (*This acknowledgement is optional.*)

I added a method to the CodonAdaptationIndex class in biopython/Bio/SeqUtils/__init__.py. This method, optimize(), accepts a DNA or protein sequence and returns a DNA sequence coding for the same amino acids, but using only the preferred codons from the CodonAdaptationIndex instance.

I added a couple of lines in biopython/Tests/test_SeqUtils.py to apply the method to a sequence and assert that the calculated CAI after optimizing is equal to 1.

### CODE REVIEW

**Comment by peterjc** [comment]
Looks like this replaced #4367 - GitHub lets you update your branch and it updates the pull request to match.

File: Bio/SeqUtils/__init__.py [hunk 1]
```diff
@@ -667,6 +667,40 @@ def calculate(self, sequence):
...
+        for codon, aminoacid in table.forward_table.items():
+            if self[codon] == 1.0:
+                pref_codons[aminoacid] = codon
+        # Create amino acid sequence if DNA was provided
+        if seq_type == "DNA":
+            aa_seq = translate(seq)
+        elif seq_type == "protein":
+            aa_seq = seq
+        else:
+            raise TypeError
```
Comment (peterjc) [inline Bio/SeqUtils/__init__.py:689]: Is it not meaningful to support RNA?

File: Bio/SeqUtils/__init__.py [hunk 2]
```diff
@@ -667,6 +667,40 @@ def calculate(self, sequence):
...
+        for aa in aa_seq:
+            try:
+                pref_codon = pref_codons[aa]
+            except KeyError:  # Keep stop codon from original sequence
+                if aa == "*":
+                    pref_codon = codon
+                    continue
+                else:
+                    print(f"Unrecognized codon: {codon}")
+            optimized += pref_codon
```
Comment (peterjc) [inline Bio/SeqUtils/__init__.py:701]: How long are these strings (and how often might a real use case call this method)?

I've not benchmarked this in years but this used to be true:

Slow:

```
my_string = ""
for letter in some_long_loop:
    my_string += letter
```

Fast:

```
my_letters = []
for letter in some_long_loop:
    my_letters.append(letter)
my_string = "".join(my_letters)
```

File: Bio/SeqUtils/__init__.py [hunk 3]
```diff
@@ -667,6 +667,40 @@ def calculate(self, sequence):
...
+                pref_codons[aminoacid] = codon
+        # Create amino acid sequence if DNA was provided
+        if seq_type == "DNA":
+            aa_seq = translate(seq)
+        elif seq_type == "protein":
+            aa_seq = seq
+        else:
+            raise TypeError
+        # Un-translate in loop using only preferred codons
+        optimized = Seq("")
```
Comment (peterjc) [inline Bio/SeqUtils/__init__.py:691]: Creating a new ``Seq`` object in each loop iteration is an easily avoided overhead.

Work with strings and at the end you can do:

```python
return Seq(optimized)
```

File: Tests/test_SeqUtils.py [hunk 4]
```diff
@@ -266,6 +266,9 @@ def test_codon_adaptation_index_initialization(self):
         )
         value = cai.calculate("ATGCGTATCGATCGCGATACGATTAGGCGGATG")
         self.assertAlmostEqual(value, 0.70246, places=5)
+        optimized_sequence = cai.optimize("ATGCGTATCGATCGCGATACGATTAGGCGGATG")
+        optimized_value = cai.calculate(optimized_sequence)
+        self.assertEqual(optimized_value, 1.0)
```
Comment (peterjc) [inline Tests/test_SeqUtils.py:271]: Can you add a test where the optimiszed value is not one?

**Review by peterjc — APPROVED** [review]
I've not looked at the details of codon optimisation scoring, but found several generic points that I'd like you to address.

Hopefully someone else can look at the core calculation...

File: Bio/SeqUtils/__init__.py (hunk 3 above)
Comment (mdehoon) [reply Bio/SeqUtils/__init__.py:691]: > Work with strings

or better yet, `bytes` objects; a string would require an extra encoding step.

**Review by peterjc — CHANGES_REQUESTED** [review]
(Sorry - I clicked the wrong button - should be changes requested, not approved as it is)

File: Tests/test_SeqUtils.py (hunk 4 above)
Comment (crockeraw) [reply Tests/test_SeqUtils.py:271]: Thanks for feedback! The value for an optimized sequence should always be 1. I could test that the optimized sequence still codes for the same amino acids as the original sequence, though. I could also do a test using amino acid sequence as the input; I have used that option less in practice.

File: Bio/SeqUtils/__init__.py (hunk 2 above)
Comment (crockeraw) [reply Bio/SeqUtils/__init__.py:701]: I imagine this function is only useful for coding sequences, since anything else would not use codons. Coding sequences are usually only a couple thousand base pairs in my experience. They do occasionally get very large though (low millions). I'll benchmark on a very large sequence and get back to you.

File: Tests/test_SeqUtils.py (hunk 4 above)
Comment (peterjc) [reply Tests/test_SeqUtils.py:271]: You can see I've not looked at the caculation details ;)

File: Bio/SeqUtils/__init__.py (hunk 2 above)
Comment (crockeraw) [reply Bio/SeqUtils/__init__.py:701]: Timeit results from optimizing titin, one of the largest proteins in the human genome:
Using str: 145 ms ± 18.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
Using list: 3.3 ms ± 322 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

File: Bio/SeqUtils/__init__.py (hunk 1 above)
Comment (crockeraw) [reply Bio/SeqUtils/__init__.py:691]: I didn't imagine a case where I would want to use RNA, but it's possible someone would. 
DNA and RNA would be handled the same way by translate(), though; I've update the code to accept seq_type = "RNA".

File: Bio/SeqUtils/__init__.py (hunk 3 above)
Comment (crockeraw) [reply Bio/SeqUtils/__init__.py]: I have changed `optimized` to an empty list which is populated inside the loop and then transformed into a `seq` object before being returned. Does this resolve the problem pointed out here?

File: Bio/SeqUtils/__init__.py (hunk 2 above)
Comment (peterjc) [reply Bio/SeqUtils/__init__.py]: Thanks - good to know that coding approach is still relevant on modern Python.

File: Bio/SeqUtils/__init__.py (hunk 3 above)
Comment (peterjc) [reply Bio/SeqUtils/__init__.py]: This looks good. Michiel is probably right that using bytes would be even faster - provided the subsequences you are using are already available as bytes.

**Review by peterjc — APPROVED** [review]
Looks OK to me (without understanding the core algorithm being implemented)

**Comment by peterjc** [comment]
GitHub says there is a conflict to resolve - not sure where, we might have to fix that at the command line if the GitHub interface won't let me do it here.

**Comment by peterjc** [comment]
Ah never mind - we can do a squash-and-merge, meaning all your changes would become a single commit.

GitHub is saying we can't do a rebase-and-merge, probably because your branch history has several merges in it.

**Comment by crockeraw** [comment]
When I have used this method, I use a CAI generated from an organism's entire .cds file. In this case it is very unlikely that any two codons would be equally preferred (both have value 1). If someone did use the `optimize` method with multiple equally-preferred codons (for instance if they were relying only on a ribosomal protein sequence) , it would only use the one which came later in `standard_dna_table.forward_table.items()`.

One solution to this issue could be to provide a warning that multiple codons are equivalently preferred, and a message saying which one is being (arbitrarily) used for the optimized DNA sequence. Alternatively, it could refuse to run. Or alternate between the equally-preferred codons. Thoughts?

**Comment by peterjc** [comment]
Can you make a test case for that corner case? My inclination would be pick something arbitrary but consistent fir a tie-break (e.g. alphabetical sort) and issue a warning (``import warnings`` etc).

A random choice is also sensible, it makes the test case harder ever so slightly harder to write ;)

**Comment by crockeraw** [comment]
I handled it with a warning. The warning indicates which codon is used. A test case actually already does include an amino acid (S) with two equally-preferred codons (TCA and AGT) and passes just fine. If a user wanted to force one codon to be used over another, I think altering the CAI to a smaller number (like .99) for the codon they did not want included would be an obvious solution.

File: Bio/SeqUtils/__init__.py [hunk 5]
```diff
@@ -667,6 +667,46 @@ def calculate(self, sequence):
...
 
+    def optimize(self, sequence, seq_type="DNA", table=standard_dna_table):
+        """Return a new DNA sequence with preferred codons only."""
+        try:  # If seq record is provided, convert to sequence
+            sequence = sequence.seq
+        except AttributeError:  # not a  SeqRecord object
+            pass
+        seq = sequence.upper()
+        # Make dict with amino acids referencing preferred codons
+        pref_codons = {aminoacid: [] for aminoacid in table.protein_alphabet}
```
Comment (speleo3) [inline Bio/SeqUtils/__init__.py:678]: Using an empty list as a placeholder for unset items is odd. How about starting with an empty dictionary? And then in line 681 use `if aminoacid in pref_codons:` instead of `if pref_codons[aminoacid] != []:`

File: Bio/SeqUtils/__init__.py [hunk 6]
```diff
@@ -667,6 +667,46 @@ def calculate(self, sequence):
...
+        else:
+            raise TypeError
+        # Un-translate in loop using only preferred codons
+        optimized = []
+        for aa in aa_seq:
+            try:
+                pref_codon = pref_codons[aa]
+            except KeyError:  # Keep stop codon from original sequence
+                if aa == "*":
+                    pref_codon = codon
```
Comment (speleo3) [inline Bio/SeqUtils/__init__.py:702]: "Keep stop codon from original sequence"? This seems wrong. The `codon` variable you're referencing here is from the last iteration over `table.stop_codons` on line 685. Or, if there would be no stop codons in the table, then it would be from the last iteration over `table.forward_table.items()`. But, assuming there are stop codons in the table, then `"*"` can't throw a `KeyError` and  line 702 would be unreachable.

File: Bio/SeqUtils/__init__.py [hunk 7]
```diff
@@ -667,6 +667,46 @@ def calculate(self, sequence):
 
         return exp(cai_value / cai_length)
 
+    def optimize(self, sequence, seq_type="DNA", table=standard_dna_table):
```
Comment (speleo3) [inline Bio/SeqUtils/__init__.py:670]: Would it be meaningful to use a different `table` than the one already passed to `CodonAdaptationIndex.__init__`? I don't think so. Instead, I'd suggest to add `self._table = table` to `__init__` and then use that here in the `optimize` method.

File: Bio/SeqUtils/__init__.py (hunk 6 above)
Comment (crockeraw) [reply Bio/SeqUtils/__init__.py:702]: Thanks! You're right, that was something I'd meant to remove.

File: Bio/SeqUtils/__init__.py (hunk 7 above)
Comment (crockeraw) [reply Bio/SeqUtils/__init__.py:670]: Great point - will make change

File: Bio/SeqUtils/__init__.py (hunk 5 above)
Comment (crockeraw) [reply Bio/SeqUtils/__init__.py]: Right, that is more natural, I have updated as suggested. Thanks!

File: Bio/SeqUtils/__init__.py [hunk 8]
```diff
@@ -667,6 +668,42 @@ def calculate(self, sequence):
...
+        else:
+            raise TypeError
+        # Un-translate in loop using only preferred codons
+        optimized = []
+        for aa in aa_seq:
+            try:
+                pref_codon = pref_codons[aa]
+            except KeyError:
+                print(f"Unrecognized amino acid: {aa}")
+            optimized.append(pref_codon)
```
Comment (speleo3) [inline Bio/SeqUtils/__init__.py:703]: Two problems:
- first letter is unrecognized -> `UnboundLocalError`
- second letter is unrecognized -> repeats first letter

If you intent to drop those unrecognized amino acids, I suggest to add a test like this:
```py
assert cai.optimize("EDU", "protein").translate() == "ED"
assert cai.optimize("EUD", "protein").translate() == "ED"
assert cai.optimize("UED", "protein").translate() == "ED"
```

Comment (peterjc) [reply Bio/SeqUtils/__init__.py:703]: Raising an exception (ValueError? Maybe KeyError) would seem fine too - and you can write tests to confrim it raises an exception.

Comment (crockeraw) [reply Bio/SeqUtils/__init__.py:703]: Thanks, good catch. I've made changes to raise a KeyError on invalid codons and added a test to confirm a KeyError is raised.

**Review by speleo3 — APPROVED** [review]

File: Bio/SeqUtils/__init__.py (hunk 8 above)
Comment (speleo3) [reply Bio/SeqUtils/__init__.py:703]: Looks good to me now, thanks for the changes.

FYI, I would probably opt for this :-)
```py
        try:
            optimized_str = "".join(pref_codons[aa] for aa in seq)
        except KeyError as ex:
            raise KeyError(f"Unrecognized amino acid: {ex}") from None
```

**Comment by speleo3** [comment]
What's biopython's take on type hints? How about this?
```py
    def optimize(
        self,
        sequence: Seq,
        seq_type: typing.Literal["DNA", "RNA", "protein"] = "DNA",
    ) -> Seq:
```
By the way, I'm not a fan of accepting both `SeqRecord` and `Seq` for the sequence argument, but I see that it's consistent with other functions in this module.

**Comment by peterjc** [comment]
We've had a few efforts at adding type hints, and I'm all for it **IF** we can get mypy or similar included in the CI to keep them tested. See e.g. #2236 and #3165

**Comment by speleo3** [comment]
Thanks Peter, that's good to know. Too bad the pull request got stuck in 2020 :-( Adding mypy CI to a large existing codebase is a pretty high bar to pass I suppose.

**Comment by peterjc** [comment]
I was hoping we could start with typing a narrow subset of the code, but I don't know enough about mypy to say for sure if that's the best idea. Perhaps I'll use it on a side project to learn more about it...

File: Bio/SeqUtils/__init__.py (hunk 1 above)
Comment (peterjc) [reply Bio/SeqUtils/__init__.py:695]: Looking at this again, two minor points.

First can you add an exception message about the argument not being a supported sequence type.

And second, maybe ValueError not TypeError?

File: Bio/SeqUtils/__init__.py (hunk 8 above)
Comment (peterjc) [reply Bio/SeqUtils/__init__.py:703]: That looks shorter/faster, but is the exception as helpful? i.e. Does it tell you which amino acid was a problem

Comment (speleo3) [reply Bio/SeqUtils/__init__.py:703]: Yes, the exception will be the same, `str(ex)` will be the unrecognized amino acid.

**Comment by speleo3** [comment]
I'll try to rebase the `mr-c:mypyc_test` branch and get it working :crossed_fingers:

**Comment by peterjc** [comment]
I need to run black style - at times like this I am reminded we could have that happen automatically on pull requests, e.g. see #4322

**Comment by mdehoon** [comment]
@crockeraw  Can you add some more explanation to the docstring? It's not clear to me from the docstring what the input and output is, and when one would use this function.

File: Bio/SeqUtils/__init__.py [hunk 9]
```diff
@@ -667,6 +668,40 @@ def calculate(self, sequence):
...
+        except AttributeError:  # not a  SeqRecord object
+            pass
+        seq = sequence.upper()
+        # Make dict with amino acids referencing preferred codons
+        pref_codons = {}
+        for codon, aminoacid in self._table.forward_table.items():
+            if self[codon] == 1.0:
+                if aminoacid in pref_codons:
+                    message = f"{pref_codons[aminoacid]} and {codon} are equally preferred. Using {codon}"
+                    warnings.warn(message, RuntimeWarning)
```
Comment (mdehoon) [inline Bio/SeqUtils/__init__.py:684]: Note that by default, the warning will be shown only once in a single Python session for a specific aminoacid and codon. Calling optimize twice with the same sequence would show the warning only the first time. Also, if a subsequent sequence contains the same codon, then again the warning will not be shown.
A better way is to have strict=True|False argument that specifies if an Exception should be raised if equally preferred codons are found.

Comment (peterjc) [reply Bio/SeqUtils/__init__.py:684]: I wouldn't object to the suggested strict mode, but personally am content with the proposed warning behaviour.

**Comment by peterjc** [comment]
+1 on @mdehoon's suggestion to expand the docstring.

Ideally this would include a (short) doctest, but given the ``CodonAdaptationIndex`` class doesn't seem to have any at all, that would be a bigger ask.

**Comment by crockeraw** [comment]
I could add a doctest for `CodonAdaptationIndex` and for each of its functions in a separate pull request if that that would be helpful and cleaner than tacking on here.

File: Bio/SeqUtils/__init__.py [hunk 10]
```diff
@@ -667,6 +668,66 @@ def calculate(self, sequence):
...
+                    if strict:
+                        raise Exception(
+                            f"{pref_codons[aminoacid]} and \
+                                        {codon} are equally preferred."
+                        )
+                    else:
+                        message = f"{pref_codons[aminoacid]} and {codon} \
+                                    are equally preferred. Using {codon}"
+                        warnings.warn(message, RuntimeWarning)
+                pref_codons[aminoacid] = codon
```
Comment (speleo3) [inline Bio/SeqUtils/__init__.py:711]: The backslash inside the string puts a lot of whitespace in the error message. I would probably do something like this:
```py
                    msg = f"{pref_codons[aminoacid]} and {codon} are equally preferred."
                    if strict:
                        raise Exception(msg)
                    warnings.warn(f"{msg} Using {codon}", RuntimeWarning)
```

**Comment by peterjc** [comment]
@crockeraw A separate follow up pull request adding doctests to the module would be more welcome.

File: Bio/SeqUtils/__init__.py (hunk 10 above)
Comment (peterjc) [reply Bio/SeqUtils/__init__.py:706]: Good catch. You can also do something like this (double check with black), using implicit string concatenation:

```python
                        message = (
                            f"{pref_codons[aminoacid]} and {codon} "
                            f"are equally preferred. Using {codon}"
                        )
```

Biopython as a whole is not consistent about having a trailing space at the end, versus a leading space at the continuing line.

File: Bio/SeqUtils/__init__.py [hunk 11]
```diff
@@ -667,6 +668,61 @@ def calculate(self, sequence):
...
+        seq = sequence.upper()
+        # Make dict with amino acids referencing preferred codons
+        pref_codons = {}
+        for codon, aminoacid in self._table.forward_table.items():
+            if self[codon] == 1.0:
+                if aminoacid in pref_codons:
+                    msg = f"{pref_codons[aminoacid]} and {codon} are equally preferred."
+                    if strict:
+                        raise Exception(msg)
+                    warnings.warn(f"{msg} Using {codon}", RuntimeWarning)
```
Comment (mdehoon) [inline Bio/SeqUtils/__init__.py:705]: Do you still need this warning?

File: Bio/SeqUtils/__init__.py [hunk 12]
```diff
@@ -667,6 +668,61 @@ def calculate(self, sequence):
...
+            pass
+        seq = sequence.upper()
+        # Make dict with amino acids referencing preferred codons
+        pref_codons = {}
+        for codon, aminoacid in self._table.forward_table.items():
+            if self[codon] == 1.0:
+                if aminoacid in pref_codons:
+                    msg = f"{pref_codons[aminoacid]} and {codon} are equally preferred."
+                    if strict:
+                        raise Exception(msg)
```
Comment (mdehoon) [inline Bio/SeqUtils/__init__.py:704]: `ValueError` is more appropriate here than a general `Exception`

File: Bio/SeqUtils/__init__.py (hunk 11 above)
Comment (crockeraw) [reply Bio/SeqUtils/__init__.py:705]: The use case I imagine is evaluating codon preferences from an organisms coding sequences or ribosomal genes, then using those codon preferences to codon-optimize a DNA sequence or set of sequences for expression in the organism of interest. 

Equally preferred codons would only be likely when using a single, or a small number of sequences to generate the CAI table. If I were using this I would always want to be made aware of the fact that there are equally preferred codons and that one was used over the other in generating the optimized sequence. I can imagine that someone might want to throw an exception instead, but I do like keeping the warning as default behavior.

Comment (mdehoon) [reply Bio/SeqUtils/__init__.py:705]: The problem with warnings is that they are unreliable. By default, they get triggered only once, and many users won't be aware of this. It gives a false sense of security.

> If I were using this I would always want to be made aware of the fact that there are equally preferred codons and that one was used over the other in generating the optimized sequence.

You can do this by setting `strict=True`, and if the exception is raised, you can decide whether you want to repeat the calculation with `strict=False`. Effectively it's the same as using a warning, but this way it is reliable and reproducible.

Comment (peterjc) [reply Bio/SeqUtils/__init__.py:705]: So @mdehoon would you favour strict=True by default with an exception, and strict=False with no warnings?

**Comment by peterjc** [comment]
You'll need to update the tests for the change to the default.

**Comment by crockeraw** [comment]
Black is failing here but passing locally... Sorry, I'm not sure what is going on.

**Comment by crockeraw** [comment]
Is pre-commit not passing because of issues on the master branch? Due to #4428 
Or is this a separate issue?

**Comment by peterjc** [comment]
Yes, the blackend-docs failure is from #4428, you can ignore that.

**Review by mdehoon — APPROVED** [review]

**Comment by peterjc** [comment]
Thank you all, that's merged now.
